### PR TITLE
[WIP] make delegates the new way

### PR DESCRIPTION
**Pull request by blairconrad**
Relates to #1437

### CODE REVIEW

**Comment by blairconrad** [comment]
For discussion only. Gives me a place to reference the code if I want to look at it away from home.
Also we can yammer, but don't feel any pressure, @thomaslevesque.

File: src/FakeItEasy/Core/DefaultFakeManagerAccessor.cs [hunk 1]
```diff
@@ -22,13 +23,22 @@ public FakeManager GetFakeManager(object proxy)
             Guard.AgainstNull(proxy);
 
             FakeManager? result = this.TryGetFakeManager(proxy);
+            if (result is not null)
+            {
+                return result;
+            }
 
-            if (result is null)
+            if (proxy is Delegate { Target: not null } d)
```
Comment (blairconrad) [inline src/FakeItEasy/Core/DefaultFakeManagerAccessor.cs]: Sometimes we're passed the Delegate, and sometimes the DynamicProxy proxy. If the delegate...

Actually, I think I've double-fixed this issue. Here and in the generator. I'll see if I can simplify.

File: src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs [hunk 2]
```diff
@@ -23,212 +20,24 @@ internal static class DelegateProxyGenerator
...
+            fakeCallProcessorProvider.EnsureInitialized(proxy);
 
-            public override CompletedFakeObjectCall ToCompletedCall()
-            {
-                return new CompletedFakeObjectCall(
-                    this,
-                    this.originalArguments);
-            }
+            var a = ServiceLocator.Resolve<IFakeManagerAccessor>();
+            a.SetFakeManager(delegateProxy, a.GetFakeManager(proxy));
```
Comment (blairconrad) [inline src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs:51]: hack to establish a link between the delegate and the FakeManager.

**Comment by blairconrad** [comment]
image

File: src/FakeItEasy/Creation/ProxyInterceptor.cs
Comment (blairconrad) [inline src/FakeItEasy/Creation/ProxyInterceptor.cs:1]: I hoisted it to use from delegate. This is probably not the best ending place. At least in this form. Not that anything in the PR is best anything

File: src/FakeItEasy/Core/FakeManagerProvider.cs [hunk 3]
```diff
@@ -51,7 +51,9 @@ public IFakeCallProcessor Fetch(object proxy)
             }
         }
 
-        public void EnsureInitialized(object proxy)
+        public void EnsureInitialized(object proxy) => this.EnsureInitialized(proxy, null);
+
+        public void EnsureInitialized(object proxy, object? alias)
```
Comment (thomaslevesque) [inline src/FakeItEasy/Core/FakeManagerProvider.cs:56]: `alias` will always be a delegate, right? So why not type it as `Delegate` and name it `delegateProxy`?

**Pull request closed without merging**